### ISSUE TITLE
SEGV when next is followed by a goto to a label in the same block

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7897]
Migrated from [rt.perl.org#35214](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=35214) (status was 'resolved')

Searchable as RT35214$

**Comment by p5pRT** [comment]
### From bod@debian.org

#### Created by bod@debian.org



Segfault in the following code&#8203;:

\#\!/usr/bin/perl

for \($\_ = 1; $\_ \< 3; $\_\+\+\)
\{
&nbsp;    AGAIN&#8203;:
&nbsp;    if \($\_ == 1\) \{ next               \}
&nbsp;    if \($\_ == 2\) \{ $\_ = 3; goto AGAIN \}
\}
\_\_END\_\_

Backtrace&#8203;:

\#0  0x0014f48c in Perl\_pp\_leaveloop \(my\_perl=0x1fbfc0\) at pp\_ctl\.c&#8203;:1768
1768                \*\+\+newsp = sv\_mortalcopy\(\*SP\);
\(gdb\) bt
\#0  0x0014f48c in Perl\_pp\_leaveloop \(my\_perl=0x1fbfc0\) at pp\_ctl\.c&#8203;:1768
\#1  0x000bd1c8 in Perl\_runops\_debug \(my\_perl=0x1fbfc0\) at dump\.c&#8203;:1442
\#2  0x0002e9ac in S\_run\_body \(my\_perl=0x1fbfc0\, oldscope=1\) at perl\.c&#8203;:1921
\#3  0x0002e218 in perl\_run \(my\_perl=0x1fbfc0\) at perl\.c&#8203;:1840
\#4  0x000275e8 in main \(argc=2\, argv=0xeffff084\, env=0xeffff090\) at perlmain\.c&#8203;:86
\(gdb\) 

The problem only occurs when both the next and the goto are executed\,
and in the order given \(next followed by goto\)\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.4:

Configured by Debian Project at Tue Mar  8 20:30:23 EST 2005.

Summary of my perl5 (revision 5 version 8 subversion 4) configuration:
  Platform:
    osname=linux, osvers=2.4.27, archname=sparc-linux-thread-multi
    uname='linux londo 2.4.27 #1 sun sep 19 18:17:57 est 2004 sparc64 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=sparc-linux -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.4 -Dsitearch=/usr/local/lib/perl/5.8.4 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.4 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.3.5 (Debian 1:3.3.5-9)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=true, libperl=libperl.so.5.8.4
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC -fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.4:
    /etc/perl
    /usr/local/lib/perl/5.8.4
    /usr/local/share/perl/5.8.4
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.4:
    HOME=/home/bod
    LANG=en_AU
    LANGUAGE (unset)
    LC_COLLATE=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/bod/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:
    PERL_BADLANG (unset)
    SHELL=/usr/bin/ksh

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, May 04\, 2005 at 10&#8203;:37&#8203;:43AM \-0000\, Brendan O'Dea wrote&#8203;:
> Segfault in the following code&#8203;:
> 
> \#\!/usr/bin/perl
> 
> for \($\_ = 1; $\_ \< 3; $\_\+\+\)
> \{
>     AGAIN&#8203;:
>     if \($\_ == 1\) \{ next               \}
>     if \($\_ == 2\) \{ $\_ = 3; goto AGAIN \}
> \}

Fixed by the change below\. next \(and redo\) didn't retore the old
PL\_curcop\, so on next loop entry\, the oldcop stored in the context block
was the wrong one\, which then confused goto\.

This is demonstrated by the following code

&nbsp;    \#\!/usr/bin/perl \-w
&nbsp;    for \($\_=1; \(\($\_ == 2 && die\)\, $\_\<3\); $\_\+\+\) \{
&nbsp;    next if $\_ == 1;
&nbsp;    \}

$ perl592 /tmp/p
Died at /tmp/p line 3\.
$ \./fixedperl /tmp/p
Died at /tmp/p line 2\.
$

Dave

\-\- 
The Enterprise is captured by a vastly superior alien intelligence which
does not put them on trial\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#10


Change 24384 by davem@&#8203;davem\-splatty on 2005/05/04 14&#8203;:01&#8203;:40

&nbsp;    \[perl \#35214\] SEGV when next is followed by a goto
&nbsp;    
&nbsp;    next and redo didn't restore PL\_curcop

Affected files \.\.\.

\.\.\. //depot/perl/pp\_ctl\.c\#427 edit
\.\.\. //depot/perl/t/op/goto\.t\#26 edit

Differences \.\.\.

==== //depot/perl/pp\_ctl\.c\#427 \(text\) ====

@&#8203;@&#8203; \-1890\,6 \+1890\,7 @&#8203;@&#8203;
&nbsp;     SV \*\*mark;
 
&nbsp;     POPBLOCK\(cx\,newpm\);
\+    assert\(CxTYPE\(cx\) == CXt\_LOOP\);
&nbsp;     mark = newsp;
&nbsp;     newsp = PL\_stack\_base \+ cx\->blk\_loop\.resetsp;
 
@&#8203;@&#8203; \-2149\,6 \+2150\,7 @&#8203;@&#8203;
&nbsp;     TOPBLOCK\(cx\);
&nbsp;     if \(PL\_scopestack\_ix \< inner\)
&nbsp;     leave\_scope\(PL\_scopestack\[PL\_scopestack\_ix\]\);
\+    PL\_curcop = cx\->blk\_oldcop;
&nbsp;     return cx\->blk\_loop\.next\_op;
 \}
 
@&#8203;@&#8203; \-2176\,6 \+2178\,7 @&#8203;@&#8203;
&nbsp;     oldsave = PL\_scopestack\[PL\_scopestack\_ix \- 1\];
&nbsp;     LEAVE\_SCOPE\(oldsave\);
&nbsp;     FREETMPS;
\+    PL\_curcop = cx\->blk\_oldcop;
&nbsp;     return cx\->blk\_loop\.redo\_op;
 \}
 

==== //depot/perl/t/op/goto\.t\#26 \(xtext\) ====

@&#8203;@&#8203; \-7\,7 \+7\,7 @&#8203;@&#8203;
&nbsp;     @&#8203;INC = qw\(\. \.\./lib\);
 \}
 
\-print "1\.\.47\\n";
\+print "1\.\.49\\n";
 
 require "test\.pl";
 
@&#8203;@&#8203; \-414\,5 \+414\,27 @&#8203;@&#8203;
 sub c32039 \{ print $\_\[0\] eq 'foo' ? "" : "not "\, "ok 47 \- chained &goto\\n" \}
 a32039\(\);
 
\+\# \[perl \#35214\] next and redo re\-entered the loop with the wrong cop\,
\+\# causing a subsequent goto to crash
\+
\+\{
\+    my $r = runperl\(
\+        stderr => 1\,
\+        prog =>
\+'for \($\_=0;$\_\<3;$\_\+\+\)\{A&#8203;: if\($\_==1\)\{next\} if\($\_==2\)\{$\_\+\+;goto A\}\}print qq\(ok\)'
\+    \);
\+    $r =~ s/\\n//g;
\+    print "\# r=$r\\nnot " unless $r eq 'ok';
\+    print "ok 48 \- next and goto\\n";
\+
\+    $r = runperl\(
\+        stderr => 1\,
\+        prog =>
\+'for \($\_=0;$\_\<3;$\_\+\+\)\{A&#8203;: if\($\_==1\)\{$\_\+\+;redo\} if\($\_==2\)\{$\_\+\+;goto A\}\}print qq\(ok\)'
\+    \);
\+    $r =~ s/\\n//g;
\+    print "\# r=$r\\nnot " unless $r eq 'ok';
\+    print "ok 49 \- redo and goto\\n";
\+\}

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @smpeters

On Wed\, May 04\, 2005 at 03&#8203;:28&#8203;:47PM \+0100\, Dave Mitchell wrote&#8203;:
> On Wed\, May 04\, 2005 at 10&#8203;:37&#8203;:43AM \-0000\, Brendan O'Dea wrote&#8203;:
> > Segfault in the following code&#8203;:
> > 
> > \#\!/usr/bin/perl
> > 
> > for \($\_ = 1; $\_ \< 3; $\_\+\+\)
> > \{
> >     AGAIN&#8203;:
> >     if \($\_ == 1\) \{ next               \}
> >     if \($\_ == 2\) \{ $\_ = 3; goto AGAIN \}
> > \}
> 
> Fixed by the change below\. next \(and redo\) didn't retore the old
> PL\_curcop\, so on next loop entry\, the oldcop stored in the context block
> was the wrong one\, which then confused goto\.
> 
> This is demonstrated by the following code
> 
>     \#\!/usr/bin/perl \-w
>     for \($\_=1; \(\($\_ == 2 && die\)\, $\_\<3\); $\_\+\+\) \{
>     next if $\_ == 1;
>     \}
> 
> $ perl592 /tmp/p
> Died at /tmp/p line 3\.
> $ \./fixedperl /tmp/p
> Died at /tmp/p line 2\.
> $
> 

I'm guessing this is an unintentional side effect\, but this fix also took
care of the long\-standing issue of redo's within a for loop that I tried
to get some discussion about last week \(RT \#2049\)\.

Steve Peters
steve@&#8203;fisharerojo\.org

**Comment by p5pRT** [comment]
### From @demerphq

On 5/4/05\, Steve Peters \<steve@&#8203;fisharerojo\.org> wrote&#8203;:
> On Wed\, May 04\, 2005 at 03&#8203;:28&#8203;:47PM \+0100\, Dave Mitchell wrote&#8203;:
> > On Wed\, May 04\, 2005 at 10&#8203;:37&#8203;:43AM \-0000\, Brendan O'Dea wrote&#8203;:
> > > Segfault in the following code&#8203;:
> > >
> > > \#\!/usr/bin/perl
> > >
> > > for \($\_ = 1; $\_ \< 3; $\_\+\+\)
> > > \{
> > >     AGAIN&#8203;:
> > >     if \($\_ == 1\) \{ next               \}
> > >     if \($\_ == 2\) \{ $\_ = 3; goto AGAIN \}
> > > \}
> >
> > Fixed by the change below\. next \(and redo\) didn't retore the old
> > PL\_curcop\, so on next loop entry\, the oldcop stored in the context block
> > was the wrong one\, which then confused goto\.
> >
> > This is demonstrated by the following code
> >
> >     \#\!/usr/bin/perl \-w
> >     for \($\_=1; \(\($\_ == 2 && die\)\, $\_\<3\); $\_\+\+\) \{
> >       next if $\_ == 1;
> >     \}
> >
> > $ perl592 /tmp/p
> > Died at /tmp/p line 3\.
> > $ \./fixedperl /tmp/p
> > Died at /tmp/p line 2\.
> > $
> >
> 
> I'm guessing this is an unintentional side effect\, but this fix also took
> care of the long\-standing issue of redo's within a for loop that I tried
> to get some discussion about last week \(RT \#2049\)\.

And there was rejoicing in the streets\!

:\-\)

Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @smpeters

On Wed\, May 04\, 2005 at 10&#8203;:39&#8203;:59AM \-0500\, Steve Peters wrote&#8203;:
> On Wed\, May 04\, 2005 at 03&#8203;:28&#8203;:47PM \+0100\, Dave Mitchell wrote&#8203;:
> > On Wed\, May 04\, 2005 at 10&#8203;:37&#8203;:43AM \-0000\, Brendan O'Dea wrote&#8203;:
> > > Segfault in the following code&#8203;:
> > > 
> > > \#\!/usr/bin/perl
> > > 
> > > for \($\_ = 1; $\_ \< 3; $\_\+\+\)
> > > \{
> > >     AGAIN&#8203;:
> > >     if \($\_ == 1\) \{ next               \}
> > >     if \($\_ == 2\) \{ $\_ = 3; goto AGAIN \}
> > > \}
> > 
> > Fixed by the change below\. next \(and redo\) didn't retore the old
> > PL\_curcop\, so on next loop entry\, the oldcop stored in the context block
> > was the wrong one\, which then confused goto\.
> > 
> > This is demonstrated by the following code
> > 
> >     \#\!/usr/bin/perl \-w
> >     for \($\_=1; \(\($\_ == 2 && die\)\, $\_\<3\); $\_\+\+\) \{
> >     next if $\_ == 1;
> >     \}
> > 
> > $ perl592 /tmp/p
> > Died at /tmp/p line 3\.
> > $ \./fixedperl /tmp/p
> > Died at /tmp/p line 2\.
> > $
> > 
> 
> I'm guessing this is an unintentional side effect\, but this fix also took
> care of the long\-standing issue of redo's within a for loop that I tried
> to get some discussion about last week \(RT \#2049\)\.
> 

Ignore my previous message\.  This fix did nothing for RT \#2049\.

Steve Peters
steve@&#8203;fisharerojo\.org
